### PR TITLE
Add Cohere models

### PR DESCRIPTION
**Pull request by allainclair**
Add Cohere **command-r-plus** and **command-r** models and test cases for them.

* Any other test case to add?

- [x] Add env var COHERE_API_KEY in CircleCi

### CODE REVIEW

File: adapters/abstract_adapters/sdk_chat_adapter.py [hunk 1]
```diff
@@ -130,13 +130,15 @@ async def execute_async(
...
     ):
         params = self.get_params(llm_input, **kwargs)
+        stream = kwargs.get("stream", False)
+        client = self.get_async_client()
 
-        response = await self.get_async_client()(
-            model=self.get_model().name,
-            **params,
-        )
+        if getattr(self, "async_chat_stream", False):
```
Comment (allainclair) [inline adapters/abstract_adapters/sdk_chat_adapter.py:136]: It seems cohere async_chat_stream must be called without await.

File: adapters/provider_adapters/cohere_sdk_chat_provider_adapter.py [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+    vendor_name: str = PROVIDER_NAME
+    provider_name: str = PROVIDER_NAME
+
+
+MODELS = [
+    CohereModel(
+        name="command-r-plus",
+        cost=Cost(prompt=3.00e-6, completion=15.00e-6),
+        context_length=131_072,
+        # completion_length=1100,
```
Comment (allainclair) [inline adapters/provider_adapters/cohere_sdk_chat_provider_adapter.py:43]: Do we need a completion_length? I tested two long answers, it costed about 1100 and 900 tokens.

File: adapters/provider_adapters/cohere_sdk_chat_provider_adapter.py [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+                "message": {
+                    "role": "assistant",
+                    "content": response.text,
+                },
+                "finish_reason": FINISH_REASON_MAPPING.get(response.finish_reason),
+            }
+        ]
+
+        # return_prompt=True and response.prompt not working.
+        for choice in self.params["chat_history"][::-1]:  # Is this reversed order?
```
Comment (allainclair) [inline adapters/provider_adapters/cohere_sdk_chat_provider_adapter.py:139]: I was based on [Anthropic](https://github.com/withmartian/adapters/blob/main/adapters/provider_adapters/anthropic_sdk_chat_provider_adapter.py#L166) to do this.
I need to know the order of the choices. Cohere chat_history is from the first to the last. I am not sure how it is in OpenAI and Anthropic

Cohere chat_history:
```python
    response = co.chat(
        chat_history=[
            {"role": "USER", "message": "Who discovered gravity?"},
            {
                "role": "CHATBOT",
                "message": "The man who is widely credited with discovering gravity is Sir Isaac Newton",
            },
        ],
        message="What year was he born??",
        model="command-r-plus",
        return_prompt=True,
        # perform web search before answering the question. You can also use your own custom connector.
        # connectors=[{"id": "web-search"}],
    )
```

File: tests/adapters/provider_adapters/cassettes/test_cohere_sdk_chat_adapter/test_async_empty_content[command-r-plus].yaml [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+      - Python
+      x-fern-sdk-name:
+      - cohere
+      x-fern-sdk-version:
+      - 5.3.4
+    method: POST
+    uri: https://api.cohere.ai/v1/chat
+  response:
+    body:
+      string: '{"message":"invalid request: all elements in history must have a message."}'
```
Comment (allainclair) [inline tests/adapters/provider_adapters/cassettes/test_cohere_sdk_chat_adapter/test_async_empty_content[command-r-plus].yaml:31]: Cohere does not accept empty message for "test empty content", any suggestion on how to overcome this? Only these two tests (sync, async) are falling at the moment.

File: adapters/provider_adapters/cohere_sdk_chat_provider_adapter.py (hunk 2 above)
Comment (amydeng2000) [reply adapters/provider_adapters/cohere_sdk_chat_provider_adapter.py:43]: cohere doesn't have a completion length, so no

File: adapters/provider_adapters/cohere_sdk_chat_provider_adapter.py (hunk 3 above)
Comment (amydeng2000) [reply adapters/provider_adapters/cohere_sdk_chat_provider_adapter.py:139]: OpenAI's chat history is a list of conversation turns in the natural order of the messages (as shown in your screenshot)

File: tests/adapters/provider_adapters/cassettes/test_cohere_sdk_chat_adapter/test_async_empty_content[command-r-plus].yaml (hunk 4 above)
Comment (allainclair) [reply tests/adapters/provider_adapters/cassettes/test_cohere_sdk_chat_adapter/test_async_empty_content[command-r-plus].yaml]: Adding a " " for strings that are empty.

File: adapters/provider_adapters/cohere_sdk_chat_provider_adapter.py [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+
+    def get_params(self, llm_input: Conversation, **kwargs: Any) -> dict[str, Any]:
+        params = super().get_params(llm_input, **kwargs)
+        for message in params["messages"]:
+            message["role"] = MAP_CONVERSATION_ROLE_TO_COHERE.get(message["role"])
+            if isinstance(message["content"], list):
+                message["message"] = " ".join(
+                    content.get("text", "") for content in message["content"]
+                )
+            elif not message["content"]:  # Empty string
```
Comment (allainclair) [inline adapters/provider_adapters/cohere_sdk_chat_provider_adapter.py:109]: When the string is empty, add a string of size 1. Cohere accepts this.

File: pyproject.toml [hunk 6]
```diff
@@ -1,6 +1,6 @@
 [tool.poetry]
 name = "martian-adapters"
-version = "5.2.0"
+version = "5.3.0"
```
Comment (allainclair) [inline pyproject.toml:3]: bump version for new models

**Review by LukeSamkharadze — APPROVED** [review]

**Pull request merged**